File: static/wasm/src/cartridge.cpp

#include "cartridge.h"
#include "bus.h"
#include "mapper.h"
#include "mapper_000.h"
#include "mapper_001.h"
#include <string>

Cartridge::Cartridge(uint8_t * data, uint32_t size, Bus * bus) : Component(bus) {

    _prgChunkSize = 16384;
    _chrChunkSize = 8192;

    _prgSize = data[4];
    _chrSize = data[5];

    _mirrorType = (data[6] & 0x01) ? EMirrorType::VERTICAL : EMirrorType::HORIZONTAL;

    _prgStart = 16;
    _chrStart = _prgStart + _prgSize * _prgChunkSize;

    _prgLength = _prgSize * _prgChunkSize;
    _prgData = (uint8_t*)calloc(_prgLength, sizeof(uint8_t));

    for(int i = 0; i < _prgLength; i++) {
        _prgData[i] = data[i+_prgStart];
    }
    
    _chrLength = _chrSize * _chrChunkSize;
    _chrData = (uint8_t*)calloc(_chrLength, sizeof(uint8_t));

    for(int i = 0; i < _chrLength; i++) {
        _chrData[i] = data[i+_chrStart];
    }

    uint16_t mapperId = ((data[7] >> 4) << 4) | (data[6] >> 4);

    switch(mapperId) {
        case 0: {
            _mapper = new Mapper000(_prgSize, _chrSize);
            break;
        }
        case 1: {
            _mapper = new Mapper001(_prgSize, _chrSize);
            break;
        }
        default: {
            printf("Mapper %u not supported\n", mapperId);
            break;
        }
    }

}

void Cartridge::Tick(double delta) {

}

void Cartridge::Clock(long totalCycles) {

}

void Cartridge::Draw() {

}

uint8_t Cartridge::Read(uint16_t address, bool readOnly) {
    uint16_t mapped = address;
    uint8_t data = 0;
    if(_mapper) {
        mapped = _mapper->MapRead(address, data);
        if(mapped == 0xFFFFFFFF) {
            return data;
        }
    }
    if(address >= 0x0000 && address <= 0x1FFF) {
        return _chrData[mapped];
    }
    if(address >= 0x8000 && address <= 0xFFFF) {
        return _prgData[mapped];
    }
    return 0x1F;
}

void Cartridge::Write(uint16_t address, uint8_t value) {
    uint16_t mapped = address;
    uint8_t data = 0;
    if(_mapper) {
        mapped = _mapper->MapWrite(address, data);
        if(mapped == 0xFFFFFFFF) {
            
        }
    }

}

EMirrorType Cartridge::GetMirrorType() {
    return _mirrorType;
}

File: static/src/component/cartridge.ts
import { Mapper000 } from "../data/mapper_000";
import { Mapper } from "../data/memory_mapper";
import { Component } from "./component";
import {ROM} from "./rom"

export class Cartridge extends Component {
    
    private _constant: string;
    private _prgSize: number;
    private _chrSize: number;

    private _prgStart: number;
    private _chrStart: number;

    private _prgChunkSize: number;
    private _chrChunkSize: number;

    private _prgRom: ROM;
    private _chrRom: ROM;

    private _mapper: Mapper;
    
    constructor(data: number[]) {

        super();

        this._prgChunkSize = 16384;
        this._chrChunkSize = 8192;
        
        this._prgSize = data[4];
        this._chrSize = data[5];
        
        this._prgStart = 16;
        this._chrStart = this._prgStart + this._prgSize * this._prgChunkSize;
        
        let prgData: number[] = [];        
        let prgLength: number = this._prgChunkSize * this._prgSize;
        let prgIndex: number = 0;
        for(let i = this._prgStart; i < this._prgStart+prgLength; i++) {
            prgData[prgIndex++] = data[i];
        }
        this._prgRom = new ROM(prgLength, prgData);        

        let chrData: number[] = [];        
        let chrLength: number = this._chrChunkSize * this._chrSize;
        let chrIndex: number = 0;
        for(let i = this._chrStart; i < this._chrStart+chrLength; i++) {
            chrData[chrIndex++] = data[i];
        }
        this._chrRom = new ROM(chrLength, chrData);

        this._mapper = new Mapper000(this._prgSize, this._chrSize);

    }

    public Pulse() {

    }

    public Update() {

    }

    public Read(addr: number): number {
        let mapped = this._mapper.MapRead(addr);
        if(mapped >= 0) {
            if(addr >= 0x0000 && addr <= 0x1FFF) {
                return this._chrRom.Read(mapped);
            }
            if(addr >= 0x8000 && addr <= 0xFFFF) {
                return this._prgRom.Read(mapped);
            }
        }
        return -1;
    }

    public Write(addr: number, data: number) {
        let mapped = this._mapper.MapWrite(addr);
        if(mapped >= 0) {
            this._prgRom.Write(mapped, data);
        } 
    }

    public PrgRom(): ROM {
        return this._prgRom;
    }

    public ChrRom(): ROM {
        return this._chrRom
    }

}